apm-lambda-extension/main_test.go

```diff
@@ -502,191 +540,255 @@ func TestAPMServerRecovery(t *testing.T) {
-	assert.Contains(t, apmServerInternals.Data, `faas.coldstart_duration":{"value":500`)
-	assert.Contains(t, apmServerInternals.Data, `faas.timeout":{"value":5000}`)
-	assert.Contains(t, apmServerInternals.Data, `coldstart":true`)
-	assert.Contains(t, apmServerInternals.Data, `execution"`)
-	assert.Contains(t, apmServerInternals.Data, `id":"arn:aws:lambda:eu-central-1:627286350134:function:main_unit_test"`)
+
+	select {
+	case <-runApp(t, logsapiAddr):
+		assert.Contains(t, apmServerInternals.Data, `{"metadata":{"service":{"name":"1234_service-12a3","version":"5.1.3","environment":"staging","agent":{"name":"elastic-node","version":"3.14.0"},"framework":{"name":"Express","version":"1.2.3"},"language":{"name":"ecmascript","version":"8"},"runtime":{"name":"node","version":"8.0.0"},"node":{"configured_name":"node-123"}},"user":{"username":"bar","id":"123user","email":"bar@user.com"},"labels":{"tag0":null,"tag1":"one","tag2":2},"process":{"pid":1234,"ppid":6789,"title":"node","argv":["node","server.js"]},"system":{"architecture":"x64","hostname":"prod1.example.com","platform":"darwin","container":{"id":"container-id"},"kubernetes":{"namespace":"namespace1","node":{"name":"node-name"},"pod":{"name":"pod-name","uid":"pod-uid"}}},"cloud":{"provider":"cloud_provider","region":"cloud_region","availability_zone":"cloud_availability_zone","instance":{"id":"instance_id","name":"instance_name"},"machine":{"type":"machine_type"},"account":{"id":"account_id","name":"account_name"},"project":{"id":"project_id","name":"project_name"},"service":{"name":"lambda"}}}}`)
+		assert.Contains(t, apmServerInternals.Data, `faas.billed_duration":{"value":60`)
+		assert.Contains(t, apmServerInternals.Data, `faas.duration":{"value":59.9`)
+		assert.Contains(t, apmServerInternals.Data, `faas.coldstart_duration":{"value":500`)
+		assert.Contains(t, apmServerInternals.Data, `faas.timeout":{"value":5000}`)
+		assert.Contains(t, apmServerInternals.Data, `coldstart":true`)
+		assert.Contains(t, apmServerInternals.Data, `execution"`)
+		assert.Contains(t, apmServerInternals.Data, `id":"arn:aws:lambda:eu-central-1:627286350134:function:main_unit_test"`)
+	case <-time.After(5 * time.Second):
+		t.Fatalf("timed out waiting for app to finish")
+	}
+}
+
+func runApp(t *testing.T, logsapiAddr string) <-chan struct{} {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	app, err := app.New(
+		app.WithExtensionName("apm-lambda-extension"),
+		app.WithLambdaRuntimeAPI(os.Getenv("AWS_LAMBDA_RUNTIME_API")),
+		app.WithLogLevel("debug"),
+		app.WithLogsapiAddress(logsapiAddr),
+	)
+	require.NoError(t, err)
+
+	go func() {
+		require.NoError(t, app.Run(ctx))
+		cancel()
+	}()
+
+	return ctx.Done()
+}
+
+func randomAddr() string {
+	// we cannot return a port that is already in use or it
+	// would return an error when creating the server.
+	// The solution is to spawn a test server to get a random
+	// port and immediately close it so that we can use the port.
```

Comment:
This is racy though, as another process could also be trying to acquire an ephemeral port. If the test ends up being flaky,  we may need to have the extension listen on an ephemeral port, log it, and then extract the listening port from the logs. Or something like that.

Reply:
Agree, the window to hit the race is small enough that I think it's safe to ignore. Another solution I had in mind was to use unix sockets.